src/Persistence/AbstractQuery.php

```diff
@@ -0,0 +1,271 @@
+    }
+
+    abstract protected function initCount($alias = null): void;
+
+    public function aggregate($fx, $field, string $alias = null, bool $coalesce = false): self
+    {
+        $this->initWhere();
+        $this->initAggregate(...func_get_args());
+
+        $this->hookInitSelect('aggregate');
+
+        return $this;
+    }
+
+    abstract protected function initAggregate($fx, $field, string $alias = null, bool $coalesce = false): void;
+
+    public function field($fieldName, string $alias = null): self
+    {
+        $this->initWhere();
+        $this->initLimit();
+        $this->initOrder();
+        $this->initField(...func_get_args());
+
+        if ($this->model->loaded()) {
+            $this->whereId($this->model->id);
+        }
+
+        $this->hookInitSelect('field');
+
+        return $this;
+    }
+
+    abstract protected function initField($fieldName, string $alias = null): void;
+
+    protected function hookInitSelect($type): void
+    {
+        if (!$this->initialized) {
+            $this->model->hook(self::HOOK_INIT_SELECT, [$this, $type]);
+
+            $this->initialized = true;
+        }
+    }
+
+    public function where($fieldName, $operator = null, $value = null): self
+    {
+        $this->scope->addCondition(...func_get_args());
+
+        return $this;
+    }
+
+    public function whereId($id)
+    {
+        if (!$this->model->id_field) {
+            throw (new Exception('Unable to find record by "id" when Model::id_field is not defined.'))
+                ->addMoreInfo('id', $id);
+        }
+
+        $this->where($this->model->getField($this->model->id_field), $id);
+
+        return $this;
+    }
+
+    abstract protected function initWhere(): void;
+
+    public function order($field, $desc = null): self
+    {
+        $this->order[] = [$field, $desc];
+
+        $this->initOrder();
+
+        return $this;
+    }
+
+    abstract protected function initOrder(): void;
+
+    public function limit($limit, $offset = 0): self
+    {
+        $this->limit = [$limit, $offset];
+
+        $this->initLimit();
+
+        return $this;
+    }
+
+    abstract protected function initLimit(): void;
+
+    protected function getLimitArgs()
+    {
+        if ($this->limit) {
+            $offset = $this->limit[1] ?? 0;
+            $limit = $this->limit[0] ?? null;
+
+            if ($limit || $offset) {
+                if ($limit === null) {
+                    $limit = PHP_INT_MAX;
+                }
+
+                return [$limit, $offset ?? 0];
+            }
+        }
+    }
+
+    public function tryExecute(): iterable
+    {
+        try {
+            $result = $this->execute();
+        } catch (Exception $e) {
+            throw (new Exception('Unable to execute query', 0, $e))
+                ->addMoreInfo('query', $this->getDebug())
+                ->addMoreInfo('message', $e->getMessage());
+        }
+
+        return $result;
+    }
+
+    public function execute(): iterable
```

Comment:
execute of like delete is iterable as well?

Reply:
This I removed as it is wrong. Only results of SELECT are iterable